roachtest: schemachange/indexrollback/tpcc-1000 failed

SHA: https://github.com/cockroachdb/cockroach/commits/98ef7abf32784b8e837d18d10173ef083010ad45

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1067238&tab=buildLog

```
The test failed on master:
	test.go:703,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1067238-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		l
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		   9m29s       69            2.0            3.6   2952.8   8589.9   8589.9   8589.9 delivery
		   9m29s       69           29.9           32.2   5905.6  85899.3 103079.2 103079.2 newOrder
		   9m29s       69            3.0            3.5    671.1   2415.9   2415.9   2415.9 orderStatus
		   9m29s       69           39.8           27.7  11274.3 103079.2 103079.2 103079.2 payment
		   9m29s       69            5.0            3.5   3758.1   4160.7   4160.7   4160.7 stockLevel
		   9m30s       69            2.0            3.6   4831.8   6979.3   6979.3   6979.3 delivery
		   9m30s       69           77.3           32.3   6979.3 103079.2 103079.2 103079.2 newOrder
		   9m30s       69            4.0            3.5   3892.3   4831.8   4831.8   4831.8 orderStatus
		   9m30s       69           59.2           27.8   5100.3 103079.2 103079.2 103079.2 payment
		   9m30s       69            8.0            3.5    192.9   3489.7   3489.7   3489.7 stockLevel
		: signal: killed
	test.go:703,cluster.go:1484,tpcc.go:120,schemachange.go:375: got non-success terminal status "failed"

```

SHA: https://github.com/cockroachdb/cockroach/commits/3c80c5e06c4aa5ed25e1cc02b78037f6ec121939

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1072755&tab=buildLog

```
The test failed on release-2.1:
	test.go:703,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1072755-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		  0.0      0.0      0.0 newOrder
		  14m29s     2513            0.0            0.5      0.0      0.0      0.0      0.0 orderStatus
		  14m29s     2513            0.0            1.4      0.0      0.0      0.0      0.0 payment
		  14m29s     2513            1.0            0.4    104.9    104.9    104.9    104.9 stockLevel
		E181229 13:59:49.665582 1 workload/cli/run.go:402  error in payment: ERROR: context deadline exceeded (SQLSTATE XX000)
		  14m30s     2521            0.0            0.5      0.0      0.0      0.0      0.0 delivery
		  14m30s     2521            2.0            2.9    209.7 103079.2 103079.2 103079.2 newOrder
		  14m30s     2521            0.0            0.5      0.0      0.0      0.0      0.0 orderStatus
		  14m30s     2521            0.0            1.4      0.0      0.0      0.0      0.0 payment
		  14m30s     2521            1.0            0.4     48.2     48.2     48.2     48.2 stockLevel
		E181229 13:59:50.890384 1 workload/cli/run.go:402  error in newOrder: ERROR: context deadline exceeded (SQLSTATE XX000)
		: signal: killed
	test.go:703,cluster.go:1484,tpcc.go:120,schemachange.go:375: timeout with current status: "running"

```

SHA: https://github.com/cockroachdb/cockroach/commits/f3cf71db94327abb6a164ad67383c35a696ec7d8

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1077687&tab=buildLog

```
The test failed on release-2.1:
	test.go:696,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1077687-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		           0.4     65.0     75.5     75.5     75.5 delivery
		  12m14s     2561            0.0            2.2      0.0      0.0      0.0      0.0 newOrder
		  12m14s     2561            0.0            0.4      0.0      0.0      0.0      0.0 orderStatus
		  12m14s     2561            3.0            0.7 103079.2 103079.2 103079.2 103079.2 payment
		  12m14s     2561            1.0            0.4 103079.2 103079.2 103079.2 103079.2 stockLevel
		E190103 14:33:09.898238 1 workload/cli/run.go:402  error in payment: ERROR: context deadline exceeded (SQLSTATE XX000)
		  12m15s     2568            2.0            0.4     71.3     96.5     96.5     96.5 delivery
		  12m15s     2568            3.0            2.2  60129.5 103079.2 103079.2 103079.2 newOrder
		  12m15s     2568            2.0            0.4      8.1     15.7     15.7     15.7 orderStatus
		  12m15s     2568            1.0            0.7 103079.2 103079.2 103079.2 103079.2 payment
		  12m15s     2568            0.0            0.4      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	test.go:696,cluster.go:1484,tpcc.go:120,schemachange.go:375: timeout with current status: "running"

```

SHA: https://github.com/cockroachdb/cockroach/commits/431b1846249fd2d110706ad221504706014e8b70

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1079927&tab=buildLog

```
The test failed on release-2.1:
	test.go:696,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1079927-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		    3.0            3.4  66572.0  66572.0  66572.0  66572.0 newOrder
		  16m12s     2914            0.0            0.5      0.0      0.0      0.0      0.0 orderStatus
		  16m12s     2914            1.0            1.4 103079.2 103079.2 103079.2 103079.2 payment
		  16m12s     2914            0.0            0.5      0.0      0.0      0.0      0.0 stockLevel
		E190104 13:50:08.857769 1 workload/cli/run.go:402  error in newOrder: ERROR: context deadline exceeded (SQLSTATE XX000)
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		  16m13s     2915            0.0            0.5      0.0      0.0      0.0      0.0 delivery
		  16m13s     2915            0.0            3.4      0.0      0.0      0.0      0.0 newOrder
		  16m13s     2915            0.0            0.5      0.0      0.0      0.0      0.0 orderStatus
		  16m13s     2915            0.0            1.4      0.0      0.0      0.0      0.0 payment
		  16m13s     2915            2.0            0.5    285.2 103079.2 103079.2 103079.2 stockLevel
		: signal: killed
	test.go:696,cluster.go:1484,tpcc.go:120,schemachange.go:375: timeout with current status: "running"

```

SHA: https://github.com/cockroachdb/cockroach/commits/99064bf2797cce49e2d56d0324a3344cf2312aca

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1082456&tab=buildLog

```
The test failed on master:
	test.go:696,cluster.go:1164,tpcc.go:110,cluster.go:1490,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1082456-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		  0.0 delivery
		   2m53s       46           17.0           18.5  19327.4 103079.2 103079.2 103079.2 newOrder
		   2m53s       46            3.0            2.6   1140.9   3087.0   3087.0   3087.0 orderStatus
		   2m53s       46           11.0           13.3   8321.5  23622.3  64424.5  64424.5 payment
		   2m53s       46            3.0            2.2   1006.6  62277.0  62277.0  62277.0 stockLevel
		   2m54s       46            4.0            2.2  16106.1  16106.1  16106.1  16106.1 delivery
		   2m54s       46           42.0           18.6  14495.5 103079.2 103079.2 103079.2 newOrder
		   2m54s       46            4.0            2.6     71.3    369.1    369.1    369.1 orderStatus
		   2m54s       46           30.0           13.4   2281.7  81604.4  98784.2  98784.2 payment
		   2m54s       46            3.0            2.2    536.9    906.0    906.0    906.0 stockLevel
		E190106 14:02:21.927343 1 workload/cli/run.go:402  error in payment: ERROR: TransactionStatusError: transaction deadline exceeded (REASON_UNKNOWN) (SQLSTATE XX000)
		: signal: killed
	test.go:696,cluster.go:1511,tpcc.go:120,schemachange.go:375: got non-success terminal status "failed"

```

```
SELECT 1 executed successfully: probably a slow statement
timeout: "ALTER TABLE ident.ident ADD COLUMN IF NOT EXISTS VARYING BIT VARYING ( 367 ) [ ]". currently executing: map[ALTER TABLE ident.ident ADD COLUMN IF NOT EXISTS VARYING BIT VARYING ( 367 ) [ ]:1 ALTER TABLE ident.ident ADD SETTINGS REGPROCEDURE:1 ALTER TABLE ident.ident ADD INTERVAL REGPROC CREATE FAMILY:1 ALTER TABLE ident.ident ADD IF NOT EXISTS YEAR UUID [ 32 ] CONSTRAINT ident AS ( DEFAULT ) STORED:1 ALTER TABLE ident.ident ADD IFERROR REGPROC:1 ALTER TABLE ident.ident DROP TIMESTAMPTZ RESTRICT:1 ALTER TABLE ident.ident DROP IF EXISTS ident RESTRICT:1 ALTER TABLE ident.ident ADD COLUMN IF NOT EXISTS SMALLINT REGPROCEDURE FAMILY ident:1]
panic: statement exec timeout
```

Not a crash or a hosed cluster. Closing

SHA: https://github.com/cockroachdb/cockroach/commits/d0e2f78183d2ce0a6b803127ee80143571c9cd4f

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1086790&tab=buildLog

```
The test failed on release-2.1:
	test.go:696,cluster.go:1164,tpcc.go:110,cluster.go:1490,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1086790-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		 0.0 delivery
		  18m21s     1658            0.0            4.8      0.0      0.0      0.0      0.0 newOrder
		  18m21s     1658            0.0            0.6      0.0      0.0      0.0      0.0 orderStatus
		  18m21s     1658            5.0            2.9 103079.2 103079.2 103079.2 103079.2 payment
		  18m21s     1658            0.0            0.7      0.0      0.0      0.0      0.0 stockLevel
		E190110 13:36:56.024504 1 workload/cli/run.go:402  error in newOrder: ERROR: TransactionStatusError: transaction deadline exceeded (REASON_UNKNOWN) (SQLSTATE XX000)
		  18m22s     1659            1.0            0.6    637.5    637.5    637.5    637.5 delivery
		  18m22s     1659            3.0            4.8   6979.3  15569.3  15569.3  15569.3 newOrder
		  18m22s     1659            0.0            0.6      0.0      0.0      0.0      0.0 orderStatus
		  18m22s     1659            0.0            2.9      0.0      0.0      0.0      0.0 payment
		  18m22s     1659            0.0            0.7      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	test.go:696,cluster.go:1511,tpcc.go:120,schemachange.go:375: timeout with current status: "running"

```

SHA: https://github.com/cockroachdb/cockroach/commits/d0e2f78183d2ce0a6b803127ee80143571c9cd4f

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=schemachange/indexrollback/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1088865&tab=buildLog

```
The test failed on release-2.1:
	test.go:696,cluster.go:1164,tpcc.go:120,cluster.go:1490,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1088865-schemachange-indexrollback-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		          0.5    604.0    604.0    604.0    604.0 delivery
		  14m55s     3002            4.0            2.3  12884.9 103079.2 103079.2 103079.2 newOrder
		  14m55s     3002            0.0            0.4      0.0      0.0      0.0      0.0 orderStatus
		  14m55s     3002            4.0            0.9 103079.2 103079.2 103079.2 103079.2 payment
		  14m55s     3002            1.0            0.4  36507.2  36507.2  36507.2  36507.2 stockLevel
		E190111 13:20:42.314617 1 workload/cli/run.go:402  error in newOrder: ERROR: context deadline exceeded (SQLSTATE XX000)
		  14m56s     3005            0.0            0.5      0.0      0.0      0.0      0.0 delivery
		  14m56s     3005            1.0            2.3 103079.2 103079.2 103079.2 103079.2 newOrder
		  14m56s     3005            0.0            0.4      0.0      0.0      0.0      0.0 orderStatus
		  14m56s     3005            0.0            0.9      0.0      0.0      0.0      0.0 payment
		  14m56s     3005            0.0            0.4      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	test.go:696,cluster.go:1511,tpcc.go:130,schemachange.go:375: timeout with current status: "running"

```

--- a/pkg/cmd/roachtest/schemachange.go
+++ b/pkg/cmd/roachtest/schemachange.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/jobs"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/jobutils"
@@ -324,7 +325,7 @@ func makeIndexAddTpccTest(numNodes, warehouses int, length time.Duration) testSp
 }
 
 func registerSchemaChangeCancelIndexTPCC1000(r *registry) {
-	r.Add(makeIndexAddRollbackTpccTest(5, 1000, time.Hour*2))
+	r.Add(makeIndexAddRollbackTpccTest(5, 1000, time.Minute*30))
 }
 
 // Creates an index and job, returning the job ID and a notify channel for
@@ -332,36 +333,35 @@ func registerSchemaChangeCancelIndexTPCC1000(r *registry) {
 func createIndexAddJob(
 	ctx context.Context, c *cluster, prefix string,
 ) (int64, <-chan error, error) {
-	setup := func(db *gosql.DB) (txn *gosql.Tx, jobID int64, err error) {
-		txn, err = db.Begin()
-		if err != nil {
-			return nil, 0, err
-		}
-		if _, err = txn.Exec(`CREATE INDEX foo ON tpcc.order (o_carrier_id);`); err != nil {
-			return nil, 0, err
-		}
-		if err = txn.QueryRow(`SELECT job_id FROM [SHOW JOBS] ORDER BY created DESC LIMIT 1`).Scan(&jobID); err != nil {
-			return nil, 0, err
-		}
-
-		return txn, jobID, nil
-	}
-
 	conn := c.Conn(ctx, 1)
-
-	txn, jobID, err := setup(conn)
+	defer conn.Close()
+	oldJobID, err := jobutils.QueryRecentJobID(conn, 0)
 	if err != nil {
-		conn.Close()
 		return 0, nil, err
 	}
 
-	c.l.Printf("%s: created index add job %d\n", prefix, jobID)
-
+	// CREATE INDEX in a separate goroutine because it takes a lot of time.
 	notifyCommit := make(chan error)
 	go func() {
-		defer conn.Close()
-		notifyCommit <- txn.Commit()
+		newConn := c.Conn(ctx, 1)
+		defer newConn.Close()
+		_, err := newConn.Exec(`CREATE INDEX foo ON tpcc.order (o_carrier_id);`)
+		notifyCommit <- err
 	}()
+
+	// Find the job id for the CREATE INDEX.
+	var jobID int64
+	for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
+		jobID, err = jobutils.QueryRecentJobID(conn, 0)
+		if err != nil {
+			return 0, nil, err
+		}
+		if jobID != oldJobID {
+			break
+		}
+	}
+
+	c.l.Printf("%s: created index add job %d\n", prefix, jobID)
 
 	return jobID, notifyCommit, nil
 }
@@ -396,7 +396,7 @@ func makeIndexAddRollbackTpccTest(numNodes, warehouses int, length time.Duration
 						ctx,
 						conn,
 						createID,
-						0.25,
+						0.12,
 						retryOpts,
 					); err != nil {
 						return err
@@ -407,11 +407,11 @@ func makeIndexAddRollbackTpccTest(numNodes, warehouses int, length time.Duration
 					}
 					c.l.Printf("%s: canceled job %d\n", prefix, createID)
 
-					if schemaChangeErr := <-notifyCommit; !testutils.IsError(schemaChangeErr, "job canceled") {
-						return errors.Errorf("expected 'job canceled' error, but got %+v", schemaChangeErr)
+					if err := <-notifyCommit; !testutils.IsError(err, "job canceled") {
+						c.l.Printf("%s: canceled job %d, got: %+v\n", prefix, createID, err)
+						return errors.Errorf("expected 'job canceled' error, but got %+v", err)
 					}
 
-					c.l.Printf("%s: rollback began %d\n", prefix, createID)
 					rollbackID, err := jobutils.QueryRecentJobID(conn, 0)
 					if err != nil {
 						return err
@@ -419,8 +419,10 @@ func makeIndexAddRollbackTpccTest(numNodes, warehouses int, length time.Duration
 						return errors.Errorf("no rollback job created")
 					}
 
-					rollbackTimeoutSecs := int((gcTTL + (15 * time.Minute)).Seconds())
-					retryOpts = retry.Options{InitialBackoff: 5 * time.Second, Multiplier: 1, MaxRetries: rollbackTimeoutSecs / 5}
+					c.l.Printf("%s: rollback for %d began: %d\n", prefix, createID, rollbackID)
+
+					backoff := 30 * time.Second
+					retryOpts = retry.Options{InitialBackoff: backoff, MaxBackoff: backoff, Multiplier: 1, MaxRetries: int(length / backoff)}
 					return jobutils.WaitForStatus(ctx, conn, rollbackID, jobs.StatusSucceeded, retryOpts)
 				},
 				Duration: length,

